### PR TITLE
[Merged by Bors] - feat: the successor as a function `WithBot α → α`

### PR DESCRIPTION
**Pull request by YaelDillies**
This will be useful to do induction on a family of polynomials by the sum of the successor of their degree.

From GrowthInGroups (LeanCamCombi)

Co-authored-by: Andrew Yang

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Order/SuccPred/WithBot.lean [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/-
+Copyright (c) 2024 Yaël Dillies Andrew Yang. All rights reserved.
```
Comment (b-mehta) [inline Mathlib/Algebra/Order/SuccPred/WithBot.lean:2]:
```suggestion
Copyright (c) 2024 Yaël Dillies, Andrew Yang. All rights reserved.
```

File: Mathlib/Algebra/Order/SuccPred/WithBot.lean [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies Andrew Yang
```
Comment (b-mehta) [inline Mathlib/Algebra/Order/SuccPred/WithBot.lean:4]:
```suggestion
Authors: Yaël Dillies, Andrew Yang
```

File: Mathlib/Algebra/Order/SuccPred/WithBot.lean [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  rw [← WithBot.coe_natCast, succ_coe, Order.succ_eq_add_one]
+
+@[simp] lemma succ_zero : succ (0 : WithBot α) = 1 := by simpa using succ_natCast 0
+
+@[simp]
+lemma succ_one : succ (1 : WithBot α) = 2 := by simpa [one_add_one_eq_two] using succ_natCast 1
+
+@[simp]
+lemma succ_ofNat (n : ℕ) [n.AtLeastTwo] :
+    succ (no_index OfNat.ofNat n : WithBot α) = OfNat.ofNat n + 1 := succ_natCast n
```
Comment (b-mehta) [inline Mathlib/Algebra/Order/SuccPred/WithBot.lean:27]:
```suggestion
    succ (no_index (OfNat.ofNat n) : WithBot α) = OfNat.ofNat n + 1 := succ_natCast n
```

Comment (b-mehta) [reply Mathlib/Algebra/Order/SuccPred/WithBot.lean:27]: else this won't apply as you expect!

File: Mathlib/Order/SuccPred/WithBot.lean [hunk 4]
```diff
@@ -0,0 +1,72 @@
+/-
+Copyright (c) 2024 Yaël Dillies Andrew Yang. All rights reserved.
```
Comment (b-mehta) [inline Mathlib/Order/SuccPred/WithBot.lean:2]: as before

**Review by b-mehta — COMMENTED** [review]
Nice!

bors d+

File: Mathlib/Algebra/Order/SuccPred/WithBot.lean (hunk 3 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Order/SuccPred/WithBot.lean:27]: Oh right! I thought this was precisely the case where it would work

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**